exploration/0475-overriding-extending-namespacing.md

```diff
@@ -0,0 +1,334 @@
+However, there are many capabilities available in platform, library,
+or operating environment APIs that could be useful to developers and translators
+or which might be expected on a specific platform or in a specific programming language.
+In addition, we expect to provide support for markup and templating regimes.
+These need to be implemented using values not found in the default registry.
+
+An additional hope is that a robust ecosystem of function libraries will be created.
+A successful ecosystem will allow users to pick-and-choose or cherry-pick fuctions or
+options to use in a given development environment.
+Each function, option, or option value extension needs to work as seamlessly as possible
+with other add-ons and with the built-in functionality.
+
+To that end, we need to define how externally-authored functions appear in a _message_;
+how externally-authored function options (and their values) can be supported;
+and what, if any, effects this has on the namespace of functions and options.
+
+## Background
+
+_What context is helpful to understand this proposal?_
+
+One example of potential add-on functionality that can help readers understand this proposal
+is the use of _skeletons_ for date and number formatting.
+
+The JavaScript `Intl.DateTimeFormat` API provides for the customization of date and time values
+through the use of "option bags". For example:
+
+```js
+new Intl.DateTimeFormat("en-GB", {
+  weekday: "long",
+  year: "numeric",
+  month: "short",
+  day: "numeric",
+  hour: "numeric",
+  minute: "numeric",
+});
+```
+
+This might be expressed in MessageFormat as:
+
+```
+Today is {$today :datetime weekday=long year=numeric month=short day=numeric hour=numeric minute=numeric}
+```
+
+The ICU family of libraries provide a shorthand mechanism called _skeletons_ for accessing
+date and time format options without needing a verbose list of options.
+The same message might look like this with a skeleton:
+
+```
+Today is {$today :datetime skeleton=EEEEyMdjm}
+```
+
+Skeletons are not proposed for inclusion in the default registry
+because they are not universally available in all datetime formatting
+libraries.
+
+## Use-Cases
+
+_What use-cases do we see? Ideally, quote concrete examples._
+
+- Developers need to add options to the base functions to suit local needs.
+  For example, ICU's skeletons as part of the `:datetime` function
+  Support for this option needs to be specified for local implemented versions.
+
+- Developers want to write a function and access it from messages.
+
+- Developers want to import 3rd party formatting packages and use the package's
+  features from within messages.
+
+- Users want to import two or more formatting packages
+  and these might have the same-named functions.
+  For example, there might be both an HTML `p` and TTS `p`
+  function.
+
+- Users want to control how extensions are referenced in their messages.
+  For example, they might wish to make a long namespace name shorter.
```

Comment:
I've been thinking about these 3 use-cases, and I think there's a portability risk that we should call out.

I see two groups of users who could be in charge of controlling the namespace name:

* (a) Authors of custom functions bundled in "packages"; this is similar to how C/C++ or Java handle naming.
* (b) Developers who import those packages into their MFv2 implementation of choice; this is in line with how modern languages handle imports and namespaces, via some sort of `import ... as` mechanism.

The current solution proposed by this PR is to go with (b), and allow developers to adjust the namespace if necessary. The risk this brings is that these custom monikers chosen by developers are then hardcoded inside messages, hurting the portability of messages.

Example: If I choose `myFoo` as a namespace, this naming choice propagates down to all messages (`myFoo:number` etc.). These messages then require edits when they're migrated to another project, when codebases merge, or even when I change my mind and rename the namespace to `myBar`.

We can attempt to avoid or mitigate this risk (or accept it).

* Avoid: Go with (a), i.e. authors choose and control the namespace name. Since this is still likely to produce conflicts, a naming scheme similar to Java's fully qualified names would be recommended. Unfortunately, it means that namespaced names become domain.name.like long.

* Mitigate: Go with (b), i.e. developers importing packages can give them custom names. Embrace the fact that those custom names are then hardcoded in mesasges and make it easy for tooling to work with them and modify them.

Reply:
I agree with this callout. The challenge with (a) is that then you need a fairly sizeable namespace that can be made unique by function authors, e.g. it suggests reverse domain naming. 

There is a third path, which is basically (b) plus a registry (think: media types). I am not wild about creating a registry, but that would help with the portability and toe-stepping problem (while allowing users to control the prefix for their own space).

Since there is no library ecosystem currently, we can also create the mechanism but not flesh it out for 2.0. That is, we can reserve the syntax but come back later to define its usage. In some ways that's like @eemeli's proposal (it's just a convention), except that it is formally reserved (if you name a function with an interior `:`, it might not work later)